File: sdks/typescript-client/tests/client.test.js
import assert from "node:assert/strict";
import test from "node:test";

import { AgentBomApiError, AgentBomClient } from "../dist/index.js";

function jsonResponse(body, init = {}) {
  return new Response(JSON.stringify(body), {
    headers: { "content-type": "application/json" },
    status: 200,
    ...init,
  });
}

test("adds auth and tenant headers", async () => {
  let seen;
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com/",
    apiKey: "test-key",
    tenantId: "tenant-a",
    fetch: async (url, init) => {
      seen = { url, init };
      return jsonResponse({ status: "ok" });
    },
  });

  const response = await client.health();

  assert.equal(response.status, "ok");
  assert.equal(seen.url, "https://agent-bom.example.com/health");
  assert.equal(seen.init.headers["x-api-key"], "test-key");
  assert.equal(seen.init.headers["x-agent-bom-tenant-id"], "tenant-a");
});

test("builds exposure path query params", async () => {
  let seenUrl = "";
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    fetch: async (url) => {
      seenUrl = String(url);
      return jsonResponse({ paths: [], stats: { count: 0 } });
    },
  });

  await client.exposurePaths({ tenantId: "tenant-b", limit: 5, minRisk: 70 });

  assert.equal(
    seenUrl,
    "https://agent-bom.example.com/v1/graph/exposure-paths?tenant_id=tenant-b&limit=5&min_risk=70",
  );
});

test("posts deploy decision payload without undefined fields", async () => {
  let payload;
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    bearerToken: "token",
    tenantId: "tenant-c",
    fetch: async (_url, init) => {
      payload = JSON.parse(init.body);
      return jsonResponse({ decision: "allow", reasons: [] });
    },
  });

  const decision = await client.shouldIDeploy({ candidate: "flask@2.0.0" });

  assert.equal(decision.decision, "allow");
  assert.deepEqual(payload, {
    candidate: "flask@2.0.0",
    tenant_id: "tenant-c",
  });
});

test("posts normalized bulk findings", async () => {
  let seenUrl = "";
  let payload;
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    tenantId: "tenant-d",
    fetch: async (url, init) => {
      seenUrl = String(url);
      payload = JSON.parse(init.body);
      return jsonResponse({
        schema_version: "v1",
        batch_id: "batch-1",
        ingested: 1,
        tenant_total: 1,
        tenant_id: "tenant-d",
        source: "agent-runtime",
      });
    },
  });

  const response = await client.ingestFindings({
    source: "agent-runtime",
    findings: [{ id: "finding-1", severity: "high" }],
  });

  assert.equal(seenUrl, "https://agent-bom.example.com/v1/findings/bulk");
  assert.equal(response.ingested, 1);
  assert.deepEqual(payload, {
    findings: [{ id: "finding-1", severity: "high" }],
    source: "agent-runtime",
    tenant_id: "tenant-d",
  });
});

test("posts bulk findings with lifecycle fields and idempotency header", async () => {
  let payload;
  let headers;
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    tenantId: "tenant-d",
    fetch: async (_url, init) => {
      payload = JSON.parse(init.body);
      headers = init.headers;
      return jsonResponse({
        schema_version: "v1",
        batch_id: "batch-1",
        ingested: 1,
        tenant_total: 1,
        tenant_id: "tenant-d",
        source: "agent-runtime",
        observed_at: "2026-07-08T12:00:00Z",
        reconciled: 2,
      });
    },
  });

  const response = await client.ingestFindings({
    source: "agent-runtime",
    findings: [{ id: "finding-1", severity: "high" }],
    observedAt: "2026-07-08T12:00:00Z",
    reconcileAbsent: true,
    idempotencyKey: "scan-batch-42",
  });

  assert.equal(response.reconciled, 2);
  assert.equal(headers["Idempotency-Key"], "scan-batch-42");
  assert.deepEqual(payload, {
    findings: [{ id: "finding-1", severity: "high" }],
    source: "agent-runtime",
    tenant_id: "tenant-d",
    observed_at: "2026-07-08T12:00:00Z",
    reconcile_absent: true,
  });
});

test("registers dataset versions", async () => {
  let seenUrl = "";
  let payload;
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    tenantId: "tenant-d",
    fetch: async (url, init) => {
      seenUrl = String(url);
      payload = JSON.parse(init.body);
      return jsonResponse({
        schema_version: "v1",
        dataset: {
          tenant_id: "tenant-d",
          dataset_id: "hf-corpus",
          version_id: "v1",
          created_at: "2026-05-17T00:00:00Z",
          source: "ci",
        },
      });
    },
  });

  const response = await client.registerDatasetVersion({
    datasetId: "hf-corpus",
    versionId: "v1",
    source: "ci",
  });

  assert.equal(seenUrl, "https://agent-bom.example.com/v1/datasets/hf-corpus/versions");
  assert.equal(response.dataset.version_id, "v1");
  assert.deepEqual(payload, {
    version_id: "v1",
    source: "ci",
    tenant_id: "tenant-d",
  });
});

test("lists dataset versions", async () => {
  let seenUrl = "";
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    fetch: async (url) => {
      seenUrl = String(url);
      return jsonResponse({
        schema_version: "v1",
        tenant_id: "tenant-d",
        dataset_id: "hf-corpus",
        versions: [],
        count: 0,
      });
    },
  });

  const response = await client.datasetVersions("hf-corpus");

  assert.equal(seenUrl, "https://agent-bom.example.com/v1/datasets/hf-corpus/versions");
  assert.equal(response.count, 0);
});

test("lists findings with default query params", async () => {
  let seenUrl = "";
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    fetch: async (url) => {
      seenUrl = String(url);
      return jsonResponse({ findings: [], count: 0 });
    },
  });

  const response = await client.listFindings({ severity: "high" });

  assert.equal(
    seenUrl,
    "https://agent-bom.example.com/v1/findings?sort=effective_reach&limit=500&offset=0&severity=high",
  );
  assert.equal(response.count, 0);
});

test("gets a single dataset version", async () => {
  let seenUrl = "";
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    fetch: async (url) => {
      seenUrl = String(url);
      return jsonResponse({
        schema_version: "v1",
        dataset: {
          tenant_id: "tenant-d",
          dataset_id: "hf/corpus",
          version_id: "2026/05/25",
          created_at: "2026-05-25T00:00:00Z",
          source: "ci",
        },
      });
    },
  });

  const response = await client.datasetVersion("hf/corpus", "2026/05/25");

  assert.equal(
    seenUrl,
    "https://agent-bom.example.com/v1/datasets/hf%2Fcorpus/versions/2026%2F05%2F25",
  );
  assert.equal(response.dataset.version_id, "2026/05/25");
});

test("registers and reads evaluation runs", async () => {
  const seen = [];
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    tenantId: "tenant-f",
    fetch: async (url, init = {}) => {
      seen.push({ url: String(url), body: init.body ? JSON.parse(init.body) : undefined });
      return jsonResponse({
        schema_version: "evals.runs.v1",
        evaluation: {
          tenant_id: "tenant-f",
          evaluation_id: "eval-a",
          created_at: "2026-05-25T00:00:00Z",
          updated_at: "2026-05-25T00:00:00Z",
          status: "completed",
          source: "ci",
        },
        evaluations: [],
        count: 0,
        limit: 10,
        offset: 0,
      });
    },
  });

  await client.registerEvaluationRun({
    evaluationId: "eval-a",
    datasetId: "dataset-a",
    datasetVersionId: "v1",
    scores: { safety: 1 },
    source: "ci",
  });
  await client.evaluationRuns({ datasetId: "dataset-a", limit: 10 });
  await client.evaluationRun("eval-a");

  assert.deepEqual(seen, [
    {
      url: "https://agent-bom.example.com/v1/evaluations",
      body: {
        evaluation_id: "eval-a",
        dataset_id: "dataset-a",
        dataset_version_id: "v1",
        source: "ci",
        scores: { safety: 1 },
        tenant_id: "tenant-f",
      },
    },
    {
      url: "https://agent-bom.example.com/v1/evaluations?dataset_id=dataset-a&limit=10",
      body: undefined,
    },
    {
      url: "https://agent-bom.example.com/v1/evaluations/eval-a",
      body: undefined,
    },
  ]);
});

test("reads manifest and runtime index with tenant query", async () => {
  const seenUrls = [];
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    tenantId: "tenant-e",
    fetch: async (url) => {
      seenUrls.push(String(url));
      return jsonResponse({ schema_version: "v1" });
    },
  });

  await client.agentManifest();
  await client.runtimeProductionIndex();

  assert.deepEqual(seenUrls, [
    "https://agent-bom.example.com/v1/agent-bom/manifest?tenant_id=tenant-e",
    "https://agent-bom.example.com/v1/runtime/production-index?tenant_id=tenant-e",
  ]);
});

test("ingests events and reads runtime sessions and observations", async () => {
  const seen = [];
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    tenantId: "tenant-r",
    fetch: async (url, init = {}) => {
      seen.push({ url: String(url), body: init.body ? JSON.parse(init.body) : undefined });
      return jsonResponse({ schema_version: "runtime.v1" });
    },
  });

  await client.ingestRuntimeEvents({ events: [{ kind: "tool_call" }] });
  await client.runtimeSessions({ limit: 10, offset: 0 });
  await client.runtimeObservations({ sessionId: "sess-1", limit: 5 });
  await client.runtimeSessionObservations("sess/1", { limit: 5 });

  assert.deepEqual(seen, [
    {
      url: "https://agent-bom.example.com/v1/runtime/events",
      body: { events: [{ kind: "tool_call" }], tenant_id: "tenant-r" },
    },
    {
      url: "https://agent-bom.example.com/v1/runtime/sessions?tenant_id=tenant-r&limit=10&offset=0",
      body: undefined,
    },
    {
      url: "https://agent-bom.example.com/v1/runtime/observations?tenant_id=tenant-r&session_id=sess-1&limit=5",
      body: undefined,
    },
    {
      url: "https://agent-bom.example.com/v1/runtime/sessions/sess%2F1/observations?tenant_id=tenant-r&limit=5",
      body: undefined,
    },
  ]);
});

test("wraps intel lookup match and sources", async () => {
  const seen = [];
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    fetch: async (url, init = {}) => {
      seen.push({ url: String(url), body: init.body ? JSON.parse(init.body) : undefined });
      return jsonResponse({ schema_version: "intel.lookup.v1", matches: [] });
    },
  });

  await client.intelLookup("CVE-2026-0001");
  await client.intelMatch({ ecosystem: "npm", name: "demo", version: "1.0.0", limit: 3 });
  await client.intelSources();

  assert.deepEqual(seen, [
    {
      url: "https://agent-bom.example.com/v1/intel/advisories/CVE-2026-0001",
      body: undefined,
    },
    {
      url: "https://agent-bom.example.com/v1/intel/match",
      body: { ecosystem: "npm", name: "demo", version: "1.0.0", limit: 3 },
    },
    {
      url: "https://agent-bom.example.com/v1/intel/sources",
      body: undefined,
    },
  ]);
});

test("throws typed errors for non-2xx responses", async () => {
  const client = new AgentBomClient({
    baseUrl: "https://agent-bom.example.com",
    fetch: async () => jsonResponse({ detail: "blocked" }, { status: 403 }),
  });

  await assert.rejects(client.health(), (error) => {
    assert.ok(error instanceof AgentBomApiError);
    assert.equal(error.status, 403);
    assert.match(error.body, /blocked/);
    return true;
  });
});

File: sdks/go/client_test.go
package agentbom

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"
)

func testClient(t *testing.T, handler http.HandlerFunc) *Client {
	t.Helper()
	server := httptest.NewServer(handler)
	t.Cleanup(server.Close)
	client, err := NewClient(Options{
		BaseURL:  server.URL + "/",
		APIKey:   "secret",
		TenantID: "tenant-a",
	})
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	return client
}

func TestClientSetsAuthHeadersAndStripsEmptyBodyFields(t *testing.T) {
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/graph/should-i-deploy" {
			t.Fatalf("path = %s", r.URL.Path)
		}
		if got := r.Header.Get("X-API-Key"); got != "secret" {
			t.Fatalf("X-API-Key = %q", got)
		}
		if got := r.Header.Get("X-Agent-Bom-Tenant-ID"); got != "tenant-a" {
			t.Fatalf("X-Agent-Bom-Tenant-ID = %q", got)
		}
		if got := r.Header.Get("Content-Type"); got != "application/json" {
			t.Fatalf("Content-Type = %q", got)
		}
		var body JSON
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("Decode body: %v", err)
		}
		if _, ok := body["context"]; ok {
			t.Fatalf("empty context should be omitted: %#v", body)
		}
		if body["tenant_id"] != "tenant-a" || body["candidate"] != "flask@2.0.0" {
			t.Fatalf("unexpected body: %#v", body)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"decision":"allow"}`))
	})

	blockRisk := 80
	result, err := client.ShouldIDeploy(context.Background(), DeployDecisionRequest{
		Candidate: "flask@2.0.0",
		BlockRisk: &blockRisk,
	})
	if err != nil {
		t.Fatalf("ShouldIDeploy: %v", err)
	}
	if result["decision"] != "allow" {
		t.Fatalf("decision = %#v", result["decision"])
	}
}

func TestClientBuildsQueryParams(t *testing.T) {
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/graph/exposure-paths" {
			t.Fatalf("path = %s", r.URL.Path)
		}
		query := r.URL.Query()
		if query.Get("tenant_id") != "tenant-a" || query.Get("limit") != "5" || query.Get("min_risk") != "70" {
			t.Fatalf("query = %s", r.URL.RawQuery)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"paths":[]}`))
	})

	limit := 5
	minRisk := 70
	if _, err := client.ExposurePaths(context.Background(), ExposurePathQuery{Limit: &limit, MinRisk: &minRisk}); err != nil {
		t.Fatalf("ExposurePaths: %v", err)
	}
}

func TestClientExposesHeadlineRoutes(t *testing.T) {
	var seen []string
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		seen = append(seen, r.Method+" "+r.URL.Path)
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"ok":true}`))
	})

	ctx := context.Background()
	if _, err := client.AgentManifest(ctx, ""); err != nil {
		t.Fatal(err)
	}
	if _, err := client.RuntimeProductionIndex(ctx, ""); err != nil {
		t.Fatal(err)
	}
	if _, err := client.IntelLookup(ctx, "GHSA-xxxx-yyyy-zzzz"); err != nil {
		t.Fatal(err)
	}
	if _, err := client.IntelMatch(ctx, IntelMatchRequest{Ecosystem: "pypi", Name: "requests", Version: "2.31.0"}); err != nil {
		t.Fatal(err)
	}
	if _, err := client.IntelSources(ctx); err != nil {
		t.Fatal(err)
	}

	want := []string{
		"GET /v1/agent-bom/manifest",
		"GET /v1/runtime/production-index",
		"GET /v1/intel/advisories/GHSA-xxxx-yyyy-zzzz",
		"POST /v1/intel/match",
		"GET /v1/intel/sources",
	}
	if len(seen) != len(want) {
		t.Fatalf("seen = %#v", seen)
	}
	for i := range want {
		if seen[i] != want[i] {
			t.Fatalf("seen[%d] = %q, want %q", i, seen[i], want[i])
		}
	}
}

func TestClientExposesFindingsAndDatasetLoop(t *testing.T) {
	var seen []string
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		seen = append(seen, r.Method+" "+r.URL.Path)
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"ok":true}`))
	})

	ctx := context.Background()
	limit := 10
	if _, err := client.ListFindings(ctx, FindingQuery{Severity: "high", Limit: &limit}); err != nil {
		t.Fatal(err)
	}
	if _, err := client.IngestFindings(ctx, IngestFindingsRequest{Findings: []JSON{{"id": "finding-1", "severity": "high"}}, Source: "sdk-test"}); err != nil {
		t.Fatal(err)
	}
	if _, err := client.RegisterDatasetVersion(ctx, DatasetVersionRequest{DatasetID: "dataset-a", VersionID: "v1"}); err != nil {
		t.Fatal(err)
	}
	if _, err := client.DatasetVersions(ctx, "dataset-a"); err != nil {
		t.Fatal(err)
	}
	if _, err := client.DatasetVersion(ctx, "dataset-a", "v1"); err != nil {
		t.Fatal(err)
	}
	if _, err := client.RegisterEvaluationRun(ctx, EvaluationRunRequest{EvaluationID: "eval-a", DatasetID: "dataset-a", DatasetVersionID: "v1"}); err != nil {
		t.Fatal(err)
	}
	if _, err := client.EvaluationRuns(ctx, EvaluationRunQuery{DatasetID: "dataset-a", Limit: &limit}); err != nil {
		t.Fatal(err)
	}
	if _, err := client.EvaluationRun(ctx, "eval-a"); err != nil {
		t.Fatal(err)
	}

	want := []string{
		"GET /v1/findings",
		"POST /v1/findings/bulk",
		"POST /v1/datasets/dataset-a/versions",
		"GET /v1/datasets/dataset-a/versions",
		"GET /v1/datasets/dataset-a/versions/v1",
		"POST /v1/evaluations",
		"GET /v1/evaluations",
		"GET /v1/evaluations/eval-a",
	}
	if len(seen) != len(want) {
		t.Fatalf("seen = %#v", seen)
	}
	for i := range want {
		if seen[i] != want[i] {
			t.Fatalf("seen[%d] = %q, want %q", i, seen[i], want[i])
		}
	}
}

func TestClientRejectsAmbiguousAuth(t *testing.T) {
	_, err := NewClient(Options{BaseURL: "https://agent-bom.example.com", APIKey: "a", BearerToken: "b"})
	if err == nil {
		t.Fatal("expected ambiguous auth error")
	}
}

func TestClientRaisesAPIErrorWithBody(t *testing.T) {
	client := testClient(t, func(w http.ResponseWriter, _ *http.Request) {
		http.Error(w, `{"detail":"forbidden"}`, http.StatusForbidden)
	})

	_, err := client.Health(context.Background())
	var apiErr *APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("expected APIError, got %T: %v", err, err)
	}
	if apiErr.StatusCode != http.StatusForbidden || apiErr.Body == "" {
		t.Fatalf("unexpected APIError: %#v", apiErr)
	}
}

func TestClientRuntimeEventsAndSessions(t *testing.T) {
	var seen []string
	client := testClient(t, func(w http.ResponseWriter, r *http.Request) {
		seen = append(seen, r.Method+" "+r.URL.RequestURI())
		if r.URL.Path == "/v1/runtime/events" {
			var body JSON
			if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
				t.Fatalf("Decode body: %v", err)
			}
			if body["tenant_id"] != "tenant-a" {
				t.Fatalf("tenant_id = %#v", body["tenant_id"])
			}
			if _, ok := body["events"]; !ok {
				t.Fatalf("events missing: %#v", body)
			}
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"schema_version":"runtime.v1"}`))
	})

	ctx := context.Background()
	limit := 5
	if _, err := client.IngestRuntimeEvents(ctx, IngestRuntimeEventsRequest{Events: []JSON{{"kind": "tool_call"}}}); err != nil {
		t.Fatalf("IngestRuntimeEvents: %v", err)
	}
	if _, err := client.RuntimeSessions(ctx, RuntimeSessionQuery{Limit: &limit}); err != nil {
		t.Fatalf("RuntimeSessions: %v", err)
	}
	if _, err := client.RuntimeObservations(ctx, RuntimeObservationQuery{SessionID: "sess-1", Limit: &limit}); err != nil {
		t.Fatalf("RuntimeObservations: %v", err)
	}
	if _, err := client.RuntimeSessionObservations(ctx, "sess/1", RuntimeSessionObservationQuery{Limit: &limit}); err != nil {
		t.Fatalf("RuntimeSessionObservations: %v", err)
	}

	want := []string{
		"POST /v1/runtime/events",
		"GET /v1/runtime/sessions?limit=5&tenant_id=tenant-a",
		"GET /v1/runtime/observations?limit=5&session_id=sess-1&tenant_id=tenant-a",
		"GET /v1/runtime/sessions/sess%2F1/observations?limit=5&tenant_id=tenant-a",
	}
	if len(seen) != len(want) {
		t.Fatalf("seen = %#v", seen)
	}
	for i := range want {
		if seen[i] != want[i] {
			t.Fatalf("request %d = %q, want %q", i, seen[i], want[i])
		}
	}
}
